Changed getExtensions() so extensionNames expands when contents do not match

ern2 opened the pull request:
This pull request accompanies JIRA ticket TRUNK-4850 (https://issues.openmrs.org/browse/TRUNK-4850)

In Module.java, getExtensions() was changed so that extensionNames are expanded when its contents do not match extensions.  Originally, getExtensions() would not call expandExtensionNames() if extensions and extensionNames were the same size, even if the contents of the two variables did not match.  The JIRA ticket goes into more detail on this issue.  

The Unit tests added in ModuleTest.java test these changes when extensionNames is null, is empty, matches extensions, and is the same size as extensions but does not match it.

dkayiwa commented on api/src/test/java/org/openmrs/module/ModuleTest.java:
```diff
@@ -0,0 +1,127 @@
+		ArrayList<Extension> ret = new ArrayList<Extension>(mockModule.getExtensions());
+
+		//getExtensionNames will return an empty extensions list since MockExtension won't
+		//have a valid class loader
```
According to this comment, this assertion has nothing to do with what this test. In other words, the assertion would pass if MockExtension had a valid class loader.

dkayiwa commented:
@ern2 can you just spy on expandExtensionNames() being called?
Here are some examples:
https://github.com/openmrs/openmrs-module-owa/commit/fdf1d06e2f19e591e2377650feb388cb1eb4d2f7
https://github.com/openmrs/openmrs-module-webservices.rest/commit/21d868e9f435f649cb2681661fa08b68790a574e

ern2 commented:
Thanks for pointing out those examples.  I will get back to you with those changes.

ern2 commented:
I updated this branch with master before this most recent commit and it now shows the file changes from the update.  Was that the wrong thing to do?  

Otherwise, I have added Powermock to my tests to spy on expandExtensionNames.

dkayiwa commented:
@ern2 this does not look correct because it includes more than your changes. Which commands did you use? By the way, have you ever seen this? https://wiki.openmrs.org/display/docs/Using+Git

ern2 commented:
@dkayiwa, I was fetching the upstream then merging instead of using rebase as is done in the wiki page you just linked.  I have re-done it now and have run mvn clean package.  I am about to push again.  Hopefully this will fix the issue.

ern2 commented:
@dkayiwa I have only made it worse.  Would it be easier at this point to delete this branch, re-branch from master, and then add my changes back in?

ern2 commented:
@dkayiwa, deleting the branch and starting over is what finally worked.  Sorry for making such a mess.  Only my changes are showing now.
